### PR TITLE
title: Add import and export subcommand for kosmosctl

### LINKED ISSUE #100: Add import  and export subcommand for kosmosctl
**Task description**: add import  and export subcommand for  kosmosctl 

**Solution**: add import  and export subcommand for  kosmosctl

### PR DESCRIPTION
**Pull request by GreatLazyMan**
Description: Add import and export subcommand for kosmosctl

#### What type of PR is this?

- **feat:** new feature

#### What does this PR do?

##export
#if svc elasticsearch-logging-data in master k8s, create ServiceExport elasticsearch-logging-data in master k8s
kosmosctl export service elasticsearch-logging-data -n foo
#if svc elasticsearch-logging-data-1 in knode knode-worker1, create ServiceExport elasticsearch-logging-data in master k8s
kosmosctl export service elasticsearch-logging-data-1 -n foo --kubeconfig /tmp/master --knode knode-worker1

##import
#if svc elasticsearch-logging-data in master k8s, create ServiceImport elasticsearch-logging-data in master k8s
kosmosctl import service elasticsearch-logging-data -n foo
#if svc elasticsearch-logging-data-1 in knode knode-worker1, create ServiceImport elasticsearch-logging-data in knode knode-worker1
kosmosctl import service elasticsearch-logging-data-1 -n foo --kubeconfig /tmp/master --knode knode-worker1
#if svc elasticsearch-logging-data-1 in knode knode-worker1, create ServiceImport elasticsearch-logging-data-1 in knode knode-worker2
kosmosctl import service elasticsearch-logging-data-1 -n foo --kubeconfig /tmp/master --knode knode-worker1 --to-knode knode-worker2

#### Which issue(s) does this PR fix?
Fixes #100

#### Screenshots (if applicable):
image

#### Additional Comments:

If svc has exported or imported, now we don't update it
image

### CODE REVIEW

File: pkg/kosmosctl/rsmigrate/rsexport.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
+package rsmigrate
+
```
Comment (duanmengkk) [inline pkg/kosmosctl/rsmigrate/rsexport.go]: [rsmigrate/rsexport.go](https://github.com/kosmos-io/kosmos/pull/101/files#diff-758b0c2da18948a96f480373079c32421c3d6c707eec0499c660422f06a03664)   you can rename to serviceexport for more readability

File: pkg/kosmosctl/rsmigrate/options.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+func (o *CommandOptions) Complete(f ctlutil.Factory, cmd *cobra.Command) error {
+	var masterConfig *restclient.Config
+	var err error
+	// get master kubernetes clientset
+	if len(o.MasterKubeConfig) > 0 {
+		masterConfig, err = clientcmd.BuildConfigFromFlags("", o.MasterKubeConfig)
+		if err != nil {
+			return fmt.Errorf("kosmosctl join complete error, generate masterConfig failed: %s", err)
```
Comment (ONE7live) [inline pkg/kosmosctl/rsmigrate/options.go]: Maybe we should modify the log description, join—>rsmigrate/import/export?

File: pkg/kosmosctl/rsmigrate/rsimport.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+			ctlutil.CheckErr(o.Complete(f, cmd))
+			ctlutil.CheckErr(o.Validate(cmd))
+			ctlutil.CheckErr(o.Run(f, cmd, args))
+			return nil
+		},
+	}
+	cmd.Flags().StringVarP(&o.MasterKubeConfig, "kubeconfig", "", "", "kubeconfig")
+	cmd.Flags().StringVarP(&o.Namespace, "namespace", "n", "default", "namespace")
+	cmd.Flags().StringVar(&o.SrcKnodeOptions.KnodeName, "node", "", "import resource from this soure knode ")
+	cmd.Flags().StringVar(&o.DstKnodeOptions.KnodeName, "to-node", "", "import resource to this destination knode")
```
Comment (ONE7live) [inline pkg/kosmosctl/rsmigrate/rsimport.go]: node—>knode and to-node—>to-knode, what do you think?

File: pkg/kosmosctl/rsmigrate/rsexport.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+		DisableFlagsInUseLine: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctlutil.CheckErr(o.Complete(f, cmd))
+			ctlutil.CheckErr(o.Validate(cmd))
+			ctlutil.CheckErr(o.Run(cmd, args))
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVarP(&o.MasterKubeConfig, "kubeconfig", "", "", "kubeconfig")
```
Comment (ONE7live) [inline pkg/kosmosctl/rsmigrate/rsexport.go]: When I use kubeconfig params, I am not quite sure what cluster's kubeconfig I should use, it is recommended to add param descriptions.

File: pkg/kosmosctl/rsmigrate/rsexport.go (hunk 1 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/rsexport.go]: modified

File: pkg/kosmosctl/rsmigrate/options.go (hunk 2 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/options.go]: modified

File: pkg/kosmosctl/rsmigrate/rsimport.go (hunk 3 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/rsimport.go]: modified

File: pkg/kosmosctl/rsmigrate/rsexport.go (hunk 4 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/rsexport.go]: modified

File: pkg/kosmosctl/rsmigrate/common.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	if err != nil {
+		return nil, nil, fmt.Errorf("create kubernetes clientset error: %s ", err)
+	}
+	crdClient, err := kosmosversioned.NewForConfig(knodeRestconfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("create kubernetes clientset for knode crd error: %s ", err)
+	}
+
+	return k8sClient, crdClient, nil
+}
```
Comment (duanmengkk) [inline pkg/kosmosctl/rsmigrate/common.go:32]: use use utils.NewKosmosClientFromBytes to cleanup the code

File: pkg/kosmosctl/rsmigrate/common.go [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	if err != nil {
+		return fmt.Errorf("get knode error: %s", err)
+	}
+	knodeOptions.Knode = knode
+	k8sClient, crdClient, err := getClientFromKnode(knodeOptions.Knode)
+	if err != nil {
+		return fmt.Errorf("get knode clientset error: %s", err)
+	}
+	knodeOptions.KnodeNativeClient = k8sClient
+	knodeOptions.KnodeCrdClient = crdClient
```
Comment (duanmengkk) [inline pkg/kosmosctl/rsmigrate/common.go]: if crdClient can rename to kosmosClient

File: pkg/kosmosctl/rsmigrate/options.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	KnodeName         string
+	Knode             *v1alpha1.Knode
+	KnodeNativeClient kubernetes.Interface
+	KnodeCrdClient    *kosmosversioned.Clientset
+}
+
+type CommandOptions struct {
+	MasterKubeConfig string
+	MasterClient     kubernetes.Interface
+	MasterCrdClient  *kosmosversioned.Clientset //clientset operate knode releted resource
```
Comment (duanmengkk) [inline pkg/kosmosctl/rsmigrate/options.go]: Comments are typically placed above the line of code.

File: pkg/kosmosctl/rsmigrate/options.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	crdClient, err := kosmosversioned.NewForConfig(masterConfig)
+	if err != nil {
+		return fmt.Errorf("create master knode clientset error: %s ", err)
+	}
+	o.MasterCrdClient = crdClient
+	masterClient, err := kubernetes.NewForConfig(masterConfig)
+	if err != nil {
+		return fmt.Errorf("create master clientset error: %s ", err)
+	}
+	o.MasterClient = masterClient
```
Comment (duanmengkk) [inline pkg/kosmosctl/rsmigrate/options.go:57]: Use the code in util to cleanup the code

File: pkg/kosmosctl/rsmigrate/serviceexport.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctlutil "k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/util/i18n"
+	"k8s.io/kubectl/pkg/util/templates"
+	mcsv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+)
+
+var exportExample = templates.Examples(i18n.T(`
+		# Export resource to control plane data storage center from control plane by default
```
Comment (duanmengkk) [inline pkg/kosmosctl/rsmigrate/serviceexport.go]:
```suggestion
		# Export service in control plane 
```

File: pkg/kosmosctl/rsmigrate/serviceexport.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+		kosmosctl export service foo  -n namespacefoo --kubeconfig=[control plane kubeconfig]
+		# Export resource to control plane data storage center from knode
```
Comment (duanmengkk) [inline pkg/kosmosctl/rsmigrate/serviceexport.go]: Currently,we do not support it

**Comment by wuyingjun-lucky** [comment]
@GreatLazyMan please check the golint ci problem

File: pkg/kosmosctl/rsmigrate/common.go (hunk 5 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/common.go:32]: modified

File: pkg/kosmosctl/rsmigrate/common.go (hunk 6 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/common.go]: modified

File: pkg/kosmosctl/rsmigrate/options.go (hunk 7 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/options.go]: modified

File: pkg/kosmosctl/rsmigrate/options.go (hunk 8 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/options.go:57]: modified

File: pkg/kosmosctl/rsmigrate/serviceexport.go (hunk 9 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/serviceexport.go]: modified

File: pkg/kosmosctl/rsmigrate/serviceexport.go (hunk 10 above)
Comment (GreatLazyMan) [reply pkg/kosmosctl/rsmigrate/serviceexport.go]: modified,now only support two kinds of command:
`kosmosctl export service elasticsearch-logging-data -n foo # create elasticsearch-logging-data serviceexport in namespace foo of control plane`
`kosmosctl import service elasticsearch-logging-data -n foo --kubeconfig /tmp/master --to-knode knode-worker1 # create serviceimport elasticsearch-logging-data in knode-worker1 on the basis of service elasticsearch-logging-data in namespace foo of control plane`

**Comment by GreatLazyMan** [comment]
> @GreatLazyMan please check the golint ci problem

Commited new code, fix golint ci problem

**Comment by GreatLazyMan** [comment]
/cc @xyz2277 @wuyingjun-lucky 
please help check

**Pull request merged**